application/src/main/java/net/corda/v5/application/crypto/SigningService.java

```diff
@@ -41,6 +41,28 @@ public interface SigningService {
     @NotNull
     DigitalSignature.WithKeyId sign(@NotNull byte[] bytes, @NotNull PublicKey publicKey, @NotNull SignatureSpec signatureSpec);
 
+    /**
+     * Using the provided signing {@link PublicKey}, internally looks up the matching {@link PrivateKey} and signs the data.
+     *
+     * @param bytes The data to sign over using the chosen key.
+     * @param publicKey The {@link PublicKey} partner to an internally held {@link PrivateKey}, either derived from the node's
+     * primary identity, or previously generated via the freshKey method. If the {@link PublicKey} is actually
+     * a {@link CompositeKey}, the first leaf signing key hosted by the node is used.
+     * @param signatureSpec The {@link SignatureSpec} to use when producing this signature.
+     * @param context The execution context of the signing operation as a map of strings. Currently accepts the following parameters
+     * <ul>
+     *   <li>category: the type of signing key requested for additional verification</li>
+     * </ul>
+     *
+     * @return A {@link DigitalSignature.WithKeyId} representing the signed data and the {@link PublicKey} that belongs to the
+     * same {@link KeyPair} as the {@link PrivateKey} that signed the data.
+     *
+     * @throws CordaRuntimeException If the input key is not a member of {@code keys}.
+     */
+    @Suspendable
+    @NotNull
+    DigitalSignature.WithKeyId sign(@NotNull byte[] bytes, @NotNull PublicKey publicKey, @NotNull SignatureSpec signatureSpec, @NotNull Map<String, String> context);
```

Comment:
Is there a particular reason why we want the extensibility of using a `Map` here? If there is, should the allowed keys also form part of the API?

Reply:
The idea was to do that because if we later add another context parameter to the operation we would need to double the overloads and it would be better to have a generalised parameter to minimise how many changes we need to make.

Reply:
Better still to pass in an object with the required fields. That way you don't have to keep adding new versions of the operation but retain a well-defined API.

Reply:
Alright, I'll look into doing that instead.

Reply:
The context parameter was something that @alexey-kadyrov-r3 and @mnesbit put in essentially at the start of the Corda 5 design. It's intentionally more dynamic so that we can add extra content without API versioning. Whether or not that's the best idea, we're basically locked down now on this since the API is considered frozen for incompatible changes.

Reply:
So, does it already exist elsewhere? I was under the impression it was something being added under this PR.

Reply:
Discussed at length on #corda-platform-crypto. To summarise, I had failed to notice that we'd taken out all these context map<string,string> parameters a few months ago, so it would now be unprecedented. It does still exist in Avro messages but that's not really relevant to the question of good design of the API. I was also referred to https://r3-cev.atlassian.net/wiki/spaces/CB/pages/4082761773/Public+API+coding+standards but I think that's out of date, trying to confirm that now on #corda-platform. As it is I think Malachy's latest iteration  may now be good, though I'm also attracted to using a primitive string type for category on sign rather than a dataclass-like additional class.